pkg/recorder/recorder.go

```diff
@@ -0,0 +1,628 @@
+	dstMask net.IPMask
+	dstPort uint16
+	proto   u8proto.U8proto
+}
+
+type RecInfo struct {
+	ID      ID
+	CapLen  uint16
+	Filters []RecorderTuple
+}
+
+type RecMask struct {
+	users int
+	prio  int
+	mask  recorderMask
+}
+
+type recQueue struct {
+	ri  *RecInfo
+	add []*RecorderTuple
+	del []*RecorderTuple
+}
+
+type Recorder struct {
+	lock.RWMutex
+	recByID map[ID]*RecInfo
+	recMask map[string]*RecMask
+	queue   recQueue
+	ctx     context.Context
+}
+
+func NewRecorder(ctx context.Context) (*Recorder, error) {
+	rec := &Recorder{
+		recByID: map[ID]*RecInfo{},
+		recMask: map[string]*RecMask{},
+		queue: recQueue{
+			add: []*RecorderTuple{},
+			del: []*RecorderTuple{},
+		},
+		ctx: ctx,
+	}
+	if option.Config.EnableRecorder {
+		maps := []*bpf.Map{}
+		if option.Config.EnableIPv4 {
+			t := &recorder.CaptureWcard4{}
+			maps = append(maps, t.Map())
+		}
+		if option.Config.EnableIPv6 {
+			t := &recorder.CaptureWcard6{}
+			maps = append(maps, t.Map())
+		}
+		for _, m := range maps {
+			if _, err := m.OpenOrCreate(); err != nil {
+				return nil, err
+			}
+			if err := m.DeleteAll(); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return rec, nil
+}
+
+func convertTupleToMask(t RecorderTuple) recorderMask {
+	m := recorderMask{
+		srcMask: make([]byte, len(t.SrcPrefix.Mask)),
+		dstMask: make([]byte, len(t.DstPrefix.Mask)),
+	}
+	if t.SrcPort != 0 {
+		m.srcPort = 0xffff
+	}
+	if t.DstPort != 0 {
+		m.dstPort = 0xffff
+	}
+	if t.Proto != 0 {
+		m.proto = 0xff
+	}
+	copy(m.srcMask, t.SrcPrefix.Mask)
+	copy(m.dstMask, t.DstPrefix.Mask)
+	return m
+}
+
+func countMaskOnes(m recorderMask) int {
+	ones := 0
+	onesSrc, _ := m.srcMask.Size()
+	onesDst, _ := m.dstMask.Size()
+	ones += onesSrc + onesDst
+	if m.srcPort == 0xffff {
+		ones += 16
+	}
+	if m.dstPort == 0xffff {
+		ones += 16
+	}
+	if m.proto == 0xff {
+		ones += 8
+	}
+	return ones
+}
+
+func hashMask(x *recorderMask) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.srcMask.String(), x.dstMask.String(),
+		int(x.srcPort), int(x.dstPort), int(x.proto))
+}
+
+func hashTuple(x *RecorderTuple) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.SrcPrefix.String(), x.DstPrefix.String(),
+		int(x.SrcPort), int(x.DstPort), int(x.Proto))
+}
+
+func (t *RecorderTuple) isIPv4() bool {
+	_, bits := t.SrcPrefix.Mask.Size()
+	return bits == 32
+}
+
+func (m *recorderMask) isIPv4() bool {
+	_, bits := m.srcMask.Size()
+	return bits == 32
+}
+
+func (m *recorderMask) genMacroSpec() string {
+	onesSrc, _ := m.srcMask.Size()
+	onesDst, _ := m.dstMask.Size()
+
+	spec := "{"
+	if m.isIPv4() {
+		spec += fmt.Sprintf(".daddr=0x%s,", m.dstMask.String())
+		spec += fmt.Sprintf(".saddr=0x%s,", m.srcMask.String())
+	} else {
+		drune := []rune(m.dstMask.String())
+		spec += fmt.Sprintf(".daddr={.d1=0x%s,.d2=0x%s},",
+			string(drune[0:8]), string(drune[8:16]))
+		srune := []rune(m.srcMask.String())
+		spec += fmt.Sprintf(".saddr={.d1=0x%s,.d2=0x%s},",
+			string(srune[0:8]), string(srune[8:16]))
```

Comment:
I don't believe this is correct. It yields `{.daddr={.d1=0xffffffff,.d2=0xffffffff}` for a mask of `ffffffffffffffffffffffff00000000` (/96).

```suggestion
		dstMask := m.dstMask.String()
		spec += fmt.Sprintf(".daddr={.d1=0x%s,.d2=0x%s},",
			dstMask[0:16], dstMask[16:32])
		srcMask := m.srcMask.String()
		spec += fmt.Sprintf(".saddr={.d1=0x%s,.d2=0x%s},",
			srcMask[0:16], srcMask[16:32])
```

Reply:
Sigh, good catch, you are correct! I'll repush once net-next CI is through. Also, thx for the simplification!